src/mlpack/tests/random_forest_test.cpp

```diff
@@ -498,3 +498,63 @@ TEST_CASE("DifferentTreesTest", "[RandomForestTest]")
 
   REQUIRE(success == true);
 }
+
+/**
+ * Test that RandomForest::Train() when passed warmStart = True trains on top
+ * of exixting forest and adds the newly trained trees to the previously
+ * exixting forest.
+ */
+TEST_CASE("WarmStartTreesTest", "[RandomForestTest]")
+{
+  arma::mat trainingData;
+  arma::Row<size_t> trainingLabels;
+  data::DatasetInfo di;
+  MockCategoricalData(trainingData, trainingLabels, di);
+
+  // Train a random forest.
+  RandomForest<> rf(trainingData, di, trainingLabels, 5, 25 /* 25 trees */, 1,
+      1e-7, 0, MultipleRandomDimensionSelect(4));
+  
+  REQUIRE(rf.NumTrees() == 25);
+
+  rf.Train(trainingData, di, trainingLabels, 5, 20 /* 20 trees */, 1, 1e-7, 0,
+      MultipleRandomDimensionSelect(4), true /* warmStart */);
+
+  REQUIRE(rf.NumTrees() == 25 + 20);
+}
+
+/**
+ * Test that RandomForest::Train() when passed warmStart = True does not drop
+ * prediction quality on train data. Note that prediction quality may drop due
+ * to overfitting in some cases.
+ */
+TEST_CASE("WarmStartTreesPredictionsQualityTest", "[RandomForestTest]")
+{
+  arma::mat trainingData;
+  arma::Row<size_t> trainingLabels;
+  data::DatasetInfo di;
+  MockCategoricalData(trainingData, trainingLabels, di);
+
+  // Train a random forest.
+  RandomForest<> rf(trainingData, di, trainingLabels, 5, 25 /* 25 trees */, 1,
+      1e-7, 0, MultipleRandomDimensionSelect(4));
+  
+  // Get performance statistics on train data.
+  arma::Row<size_t> oldPredictions;
+  rf.Classify(trainingData, oldPredictions);
+
+  // Calculate the number of correct points.
+  size_t oldCorrect = arma::accu(oldPredictions == trainingLabels);
+
+  rf.Train(trainingData, di, trainingLabels, 5, 20 /* 20 trees */, 1, 1e-7, 0,
+      MultipleRandomDimensionSelect(4), true /* warmStart */);
+
+  // Get performance statistics on train data.
+  arma::Row<size_t> newPredictions;
+  rf.Classify(trainingData, newPredictions);
+
+  // Calculate the number of correct points.
+  size_t newCorrect = arma::accu(newPredictions == trainingLabels);
+
+  REQUIRE(newCorrect - oldCorrect >= 0);
```

Comment:
I wouldn't be surprised if the random forest has already fit the data perfectly the first run.  You may want to reduce the number of trees to 1, for instance, and then warm-start and add another many trees.

Reply:
Yes, trees are really good at cramming the data XD. I will reduce the number of old trees.